pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTInstanceOfExpression.java

```diff
@@ -41,12 +47,19 @@ public Object jjtAccept(JavaParserVisitor visitor, Object data) {
         visitor.visit(this, data);
     }
 
+    @Override
+    public ASTTypeExpression getRhs() {
+        return (ASTTypeExpression) jjtGetChild(1);
+    }
 
-    /**
-     * Gets the type against which the expression is tested.
-     */
+    /** Gets the wrapped type node. */
     public ASTType getTypeNode() {
-        return (ASTType) jjtGetChild(1);
+        return getRhs().getTypeNode();
     }
 
+    @Override
+    public String getXPathNodeName() {
+        // keep it uniform for XPath
+        return "InfixExpression";
```

Comment:
I think, that  it's a matter of keeping the API simple in both languages. The class ASTInstanceOfExpression doesn't add any API that is relevant to XPath users, it's just here to provide the method `ASTType getTypeNode()` to Java clients. I think it would be more surprising for XPath users to see that the node is *not* named `InfixExpression`, since it does not have a different API (and it also extends ASTInfixExpression).

Example use case: looking for all relational expressions (`<`, `<=`, `>`, `>=`, `instanceof`).
* Java: `infixExpr.getOperator().isRelational()`
* XPath: `//.../InfixExpression[@Operator = ("<", "<=", ">", ">=", "instanceof")]`
* XPath, if we use different names: `//.../*[self::InfixExpression[@Operator = ("<", "<=", ">", ">=")] or self::InstanceOfExpression]` -> this is more complex for no added benefit.

To be honest even in Java, the benefit this class offers is very tiny. We probably could remove the class, and represent instanceof expressions with a normal ASTInfixExpression. It would be pretty easy to just cast the RHS to ASTTypeExpression to get the type node. I think I would prefer that now.

---

Expanding a bit on this topic (probably to be moved to the wiki):

We offer both an XPath API and a Java API to explore ASTs. There is nothing technically forcing us to make those representations map 1 to 1. On the one hand, keeping them close is objectively good; it eases the learning curve, and keeps the implementation straightforward. 

On the other hand, and this is my point, I think we should keep in mind, that making exceptions to this principle is not bad per se. Even if we remove ASTInstanceOfExpression, and hence don't set a precedent for this yet, we should *not* consider that it's off limits, for any language module.

Here, in the case of InfixExpression vs InstanceOfExpression, the Java API being a bit richer may improve the life of Java clients, but with the standard Java -> XPath mapping, the XPath representation is "too rich", ie makes too many needless distinctions, and XPath clients are only negatively affected.

Conversely, there are other cases where the Java -> XPath mapping may be "lossy", ie the Java representation makes no distinction where the XPath representation would need one. The example that comes to mind is comment nodes. In the XPath data model, *elements* are distinct from *comments*, even though they both are *nodes*. In fact XPath has a specific syntax to look for comments (`//comment()`). For Java, it may be simpler to represent comment nodes as plain `Node` instances (or may not, since we anyway only do so for formal comments and ignore others for now). OTOH this restricts us from using the XPath language to its full potential in XPath rules.

The ultimate compound of these problems is our `pmd-xml` module. It would be expected that an XPath rule for language `xml` behaves exactly as defined by the XPath spec (and eg, by the code samples on https://www.w3schools.com/xml/xpath_intro.asp). Yet here, the Java -> XPath mapping completely fails to represent DOM nodes correctly. We're adding attributes that were not there (`@BeginLine`, `@Image`, etc), *comments are represented with elements named "comment"*, processing instructions are not in the tree.... the list goes on.

Reply:
I understand that it makes perfect sense, that both nodes are named "InfixExpression" for XPath. But I tend to use the designer to look at the AST, even if I write a Java rule. I would be confused to see in the debugger then a InstanceOfExpression instead of a InfixExpression. It's just about "least surprise". Finding child nodes via `findChildrenOfType(InfixExpression.class)` would still work, since InstanceOfExpression is a subclass of InfixExpression, so it's not a serious problem. But I just doesn't feel right.
If possible, I'd try to avoid such situations in order not to create two versions of the AST.

Reply:
Then maybe the designer should have a way to expose the tree as it is in Java (eg add the real type name when it's not "standard"). We wanted to add an interface that the designer should use to get how a node should be displayed in the tree (as part of removing `getImage`). This could be part of that logic.

I think it's more part of the presentation logic of the designer, which should be thought of separately from the API design. I understand though, that when possible it's better to avoid the divergence anyway.

Reply:
IMHO, I would try to keep it simple...